File: learnOpencv/Final3/c++/main.cpp
#include<opencv2/core/core.hpp>
#include<opencv2/highgui/highgui.hpp>
#include <opencv2/imgcodecs.hpp> 
#include <opencv2/opencv.hpp>
#include<iostream>

using namespace std;
using namespace cv;

int main()
{
    Mat image = imread("./54.jpg");
    Mat initial = image.clone();
    Mat red1, red2, red3, open;
    cvtColor(image, image, COLOR_BGR2HSV);
    Mat image2 = image.clone();
    imshow("image", image);
    inRange(image, Scalar(0, 50, 50), Scalar(10, 255, 255),red1);
    inRange(image2, Scalar(160, 40, 40), Scalar(180, 255, 255), red2);
    addWeighted(red1, 0.5, red2, 0.5, 0,red3);
    Mat element = getStructuringElement(MORPH_RECT,Size(15, 15));
    morphologyEx(red3, open, MORPH_OPEN, element);
    imshow("red", open);
    vector<vector<Point>> contours;
    vector<Vec4i> hierarchy;
    findContours(open, contours, hierarchy, RETR_EXTERNAL, CHAIN_APPROX_SIMPLE);
    drawContours(image, contours, -1, Scalar(0, 255, 0), 1);
    for (size_t i = 0; i < contours.size(); i++) {
        Rect boundRect = boundingRect(contours[i]);
        rectangle(initial, boundRect, Scalar(0, 255, 255), 2);
    }
    imshow("ROI", initial);
    waitKey(0);
}

File: learnOpencv/Final3/python/main.py
import cv2 as cv
import copy
import numpy as np

img = cv.imread('./54.jpg')
copy = copy.deepcopy(img)
transfer = cv.cvtColor(img, cv.COLOR_BGR2HSV)
cv.imshow('hsv', transfer)
img2 = transfer
red1 = cv.inRange(transfer, np.array([0, 50, 50]), np.array([10, 255, 255]))
red2 = cv.inRange(transfer, np.array([160, 40, 40]), np.array([180,255, 255]))
red3 = cv.addWeighted(red1, 0.5, red2, 0.5, 0)
element = np.ones((15, 15), np.uint8)
open = cv.morphologyEx(red3,cv.MORPH_OPEN, element)
cv.imshow("red", open)
contours, hierarchy = cv.findContours(open, cv.RETR_EXTERNAL, cv.CHAIN_APPROX_SIMPLE )
#cv.drawContours(copy, contours, -1, (0,255,0), 1)
x,y,w,h = cv.boundingRect(contours[0])
rectangle = cv.rectangle(copy,(x,y),(x+w,y+h),(0,255,255),2)
cv.imshow("rectangle", rectangle)
cv.waitKey(0)

